### PR TITLE
Gitrest: initial write speedup

### PR DESCRIPTION
**Pull request by znewton**
## Description

When using a remote filesystem, the overhead of filesystem operations makes shredded summaries extremely slow to read/write. We already work around this for Read by computing and storing a latest "Full Summary" single file that is used on read for most "Latest" summary reads.

The only Summary Write operation in the critical path is when creating the initial summary for a new document. This PR attempts to reduce the overhead of a shredded summary read/write by "compressing" the git tree for the initial summary into a single file (using Isomorphic git with `memfs`).

## Breaking Changes

N/A, only impacts new documents

## Reviewer Guidance

I validated the following scenarios to ensure functionality with each `enableLowIoWrite` setting ("initial", true, false) as well as forward/backwards compatibility using local R11s and my fluid app.
- Write service summary from initial summary
- Write client summary from initial summary from first client
- Write client summary from initial summary from 2nd client (no service summary in the middle)
- Write service summary after client summary(ies)
- Write client summary after service summary(ies)
- Load from service summary
- Load from initial summary
- Load from client summary

### CODE REVIEW

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 1]
```diff
@@ -68,11 +101,96 @@ function getSummaryObjectFromWholeSummaryTreeEntry(entry: WholeSummaryTreeEntry)
...
+}
+function convertGitBlobToSummaryBlob(blob: IBlob): IWholeFlatSummaryBlob {
+    return {
+        content: blob.content,
+        encoding: blob.encoding === "base64" ? "base64" : "utf-8",
+        id: blob.sha,
+        size: blob.size,
+    };
+}
+function convertFullGitTreeToFullSummaryTree(
```
Comment (znewton) [inline server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:151]: This file is where the majority of the changes are. I don't like how messy it's getting

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 2]
```diff
@@ -322,7 +515,10 @@ export class GitWholeSummaryManager {
         return blobResponse.sha;
     }
 
-    private async getShaFromTreeHandleEntry(entry: IWholeSummaryTreeHandleEntry): Promise<string> {
+    private async getShaFromTreeHandleEntry(
```
Comment (yangg-msft) [inline server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:518]: Should we use the argument repoManager as opposed to this.repoManager?

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 3]
```diff
@@ -255,15 +377,76 @@ export class GitWholeSummaryManager {
...
+        );
+        Lumberjack.info("Low-IO mode: Reading summary from memory");
+        const gitTree = await inMemoryRepoManager.getTree(inMemorySummaryTreeHandle, true /* recursive */);
+        const fullGitTree = await buildFullGitTreeFromGitTree(
+            gitTree,
+            inMemoryRepoManager,
+        );
+        Lumberjack.info("Low-IO mode: Writing summary to storage");
+        const summaryTreeHandle = this.writeSummaryTreeCore(
+            [{
```
Comment (yangg-msft) [inline server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:433]: Generate a metric to capture tenant id, doc id, content length, and processing time?

Comment (hedasilv) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:433]: Or if we want to keep this as a long, we can also add those properties in the properties parameter for Lumberjack.info().

**Review by hedasilv — COMMENTED** [review]
This is great! Thanks for working on this. I have a few questions to help me understand the overall flow.

1. In `writeContainerSummary()`, an in-memory `repoManager` is used as a way to obtain the full summary associated with an initial summary upload. Traditionally, we would have very high I/O latency because internally, GitRest would break up the initial summary payload into small Git objects, and upload each one of them individually to the chosen filesystem. The trick here is, then, to do exactly the same, but instead of using the original filesystem, have the uploads happen to an in-memory filesystem. By uploading the Git Objects to the temporary, in-memory `repoManager`, and later reading from there, we end up computing the full summary without the need to make multiple network calls. And then, once we have the full summary in hand, we write it as a single blob to Git. Is that correct?
2. What happens to that initial full summary blob on subsequent summary writes? My current understanding is that on a subsequent incremental summary write, the full blob would be read from storage and parsed so GitRest and compute the incremental diff. Once that is done, it will generate Git objects (blobs, trees, commits and refs) as usual, uploading them in a "shredded" way to the `repoManager`/storage as usual, since this is not the critical path anymore. As a result, the initial full summary blob would not be referenced again, and would have to be eventually garbage collected (once that is available). Is that correct?

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts (hunk 2 above)
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:681]: I should remove3 the repoManager param here. Using `this.repoManager` in this method is intentional, because referenced summaries exist in storage, not in memory, so we never want to accidentally use the in-memory repo for this particular method.

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 4]
```diff
@@ -68,11 +101,96 @@ function getSummaryObjectFromWholeSummaryTreeEntry(entry: WholeSummaryTreeEntry)
     throw new NetworkError(400, `Unknown entry type: ${entry.type}`);
 }
 
+interface IFullGitTree {
+    tree: ITree;
+    blobs: Record<string, IBlob>;
+}
+function containsFullGitTree(gitTree: ITree): boolean {
+    return gitTree.tree.length === 1
```
[the remark this answers is not in the document]
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts]: Will do. I'll also add some brief function signature doc-comments.

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 5]
```diff
@@ -91,67 +209,62 @@ export class GitWholeSummaryManager {
...
+
+    private async readWholeSummaryTree(
+        rawTree: ITree,
+        repoManager: IRepositoryManager,
+    ): Promise<IFullSummaryTree> {
+        const fullGitTree = await parseGitTreeContainingFullGitTree(rawTree, repoManager);
+        return convertFullGitTreeToFullSummaryTree(fullGitTree);
+    }
+
+    private async getLatestVersion(repoManager: IRepositoryManager): Promise<ISummaryVersion> {
```
[the remark this answers is not in the document]
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:341]: There is not! This is a relic of a partial implementation from earlier that was refactored away. Good catch

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 6]
```diff
@@ -255,15 +377,76 @@ export class GitWholeSummaryManager {
...
+        useLowIoWrite: boolean = false,
+    ): Promise<string> {
+        if (!useLowIoWrite) {
+            return this.writeSummaryTreeCore(
+                wholeSummaryTreeEntries,
+                this.repoManager,
+            );
+        }
+
+        Lumberjack.info("Low-IO mode: Initializing In-memory Filesystem");
```
[the remark this answers is not in the document]
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:485]: I'll probably remove _most_ of these logs before merging. Currently using them for debugging

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 7]
```diff
@@ -162,10 +275,17 @@ export class GitWholeSummaryManager {
...
+    public async writeSummary(
+        payload: IWholeSummaryPayload,
+        options?: Partial<ISummaryWriteOptions>,
+    ): Promise<IWriteSummaryInfo> {
         const writeSummaryMetric = Lumberjack.newLumberMetric(
             GitRestLumberEventName.WholeSummaryManagerWriteSummary,
             this.lumberjackProperties);
+        const writeOptions: ISummaryWriteOptions = {
+            ...options,
+            ...DefaultSummaryWriteOptions,
```
[the remark this answers is not in the document]
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:361]: You were correct. Already addressed while I was validating. Good eye, thanks!

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts (hunk 2 above)
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:695]: Removed unused param and added explanatory comments

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts (hunk 3 above)
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:654]: I removed most of these inner logs, but added additional properties to the wrapping summary write metric that will help us understand each one in our telemetry (newDocument, commit/tree sha, and enableLowIoWrite).

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 8]
```diff
@@ -68,11 +101,96 @@ function getSummaryObjectFromWholeSummaryTreeEntry(entry: WholeSummaryTreeEntry)
...
+                type: "blob",
+                id: treeEntry.sha,
+                path: treeEntry.path,
+            });
+            wholeFlatSummaryBlobs.push(
+                convertGitBlobToSummaryBlob(fullGitTree.blobs[treeEntry.sha]),
+            );
+        } else {
+            wholeFlatSummaryTreeEntries.push({
+                type: "tree",
```
[the remark this answers is not in the document]
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:251]: Looks like there is no `id` property on IWholeFlatSummaryTreeEntryTree, so it's not an option to pass it along.

File: server/gitrest/packages/gitrest/config.json [hunk 9]
```diff
@@ -20,6 +20,7 @@
             "name": "nodegit"
         },
         "persistLatestFullSummary": false,
+        "enableLowIoWrite": "initial",
```
[the remark this answers is not in the document]
Comment (znewton) [reply server/gitrest/packages/gitrest/config.json]: Unfortunately, we currently have no UTs for summaries routes in Gitrest :(

[the remark this answers is not in the document]
Comment (znewton) [reply server/gitrest/packages/gitrest/config.json]: I added it to the values.yml and configmap, though. I also added `enableRepositoryManagerMetrics` to the config.json since it was missing there, but existed in the k8s charts.

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 10]
```diff
@@ -68,11 +101,96 @@ function getSummaryObjectFromWholeSummaryTreeEntry(entry: WholeSummaryTreeEntry)
...
+    gitTree: ITree,
+    repoManager: IRepositoryManager,
+): Promise<IFullGitTree> {
+    const fullGitTreeBlob = await repoManager.getBlob(
+        gitTree.tree[0].sha
+    );
+    return JSON.parse(Buffer.from(
+        fullGitTreeBlob.content,
+        fullGitTreeBlob.encoding === "base64" ? "base64" : "utf-8",
+    ).toString("utf-8")) as IFullGitTree;
```
[the remark this answers is not in the document]
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts]: Good catch! Updated to avoid unnecessary work. I'm not sure if it's ever utf-8 when being read from Git, but good looking out.

**Review by yangg-msft — APPROVED** [review]
:shipit:

File: server/gitrest/packages/gitrest-base/src/routes/summaries.ts [hunk 11]
```diff
@@ -127,6 +128,7 @@ async function createSummary(
         repoManager,
         lumberjackProperties,
         externalWriterConfig?.enabled ?? false,
+        { enableLowIoWrite },
```
Comment (zhangxin511) [inline server/gitrest/packages/gitrest-base/src/routes/summaries.ts:131]: Just for learning, how did the `GitWholeSummaryManager` consumed passed in parameter `enableLowIoWrite` ? Didn't find any other changes using this parameter.

File: server/gitrest/packages/gitrest/config.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (hedasilv) [reply server/gitrest/packages/gitrest/config.json]: Thanks! I probably missed `enableRepositoryManagerMetrics` in the config.json in my PR earlier 😅

File: server/gitrest/packages/gitrest-base/src/routes/summaries.ts (hunk 11 above)
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/routes/summaries.ts:131]: Hey Xin, GitHub doesn't expand the Diff for the `GitWholeSummaryManager` change because it's over 600 lines changed.
image
You should be able to see the changes with this param (and the actual majority of this PR) by expanding that file diff

File: server/gitrest/packages/gitrest-base/src/routes/summaries.ts [hunk 12]
```diff
@@ -116,6 +116,7 @@ async function createSummary(
     repoManagerParams: IRepoManagerParams,
     externalWriterConfig?: IExternalWriterConfig,
     persistLatestFullSummary = false,
+    enableLowIoWrite: "initial" | boolean = false
 ): Promise<IWriteSummaryResponse | IWholeFlatSummary> {
     const lumberjackProperties = {
         ...getLumberjackBasePropertiesFromRepoManagerParams(repoManagerParams),
```
Comment (hedasilv) [inline server/gitrest/packages/gitrest-base/src/routes/summaries.ts:122]: Just realized: can we add `enableLowIoWrite` as a property here under `lumberjackProperties`? That way, it will automatically be passed on to `GitWholeSummaryManager` and might make it easier to tell the metric value/datapoints for each case apart

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 13]
```diff
@@ -101,45 +141,83 @@ function getSummaryObjectFromWholeSummaryTreeEntry(entry: WholeSummaryTreeEntry)
...
+                    true, /* parseInnerFullGitTrees */
+                    // All blobs associated with full git tree are included in the full git tree blob, and
+                    // will not exists in storage individually.
+                    false, /* retrieveBlobs */
+                    depth + 1,
+                );
+                const baseTreeEntryPath = treeEntry.path.replace(fullTreePath, "");
+                treeEntries.push(...builtFullGitTree.tree.tree.map((fullTreeEntry => ({
+                    ...fullTreeEntry,
+                    path: `${baseTreeEntryPath}${fullTreeEntry.path}`,
```
Comment (hedasilv) [inline server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:187]: There is no need for a slash "/" in between `${baseTreeEntryPath}` and `${fullTreeEntry.path}`, right?

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 14]
```diff
@@ -101,45 +141,83 @@ function getSummaryObjectFromWholeSummaryTreeEntry(entry: WholeSummaryTreeEntry)
...
+                }))));
+                const fullTreeBlobs = {
+                    ...fullTree.blobs,
+                    ...builtFullGitTree.blobs,
+                };
+                blobPs.push(...Object.values(fullTreeBlobs).map(async (blob) => blob));
```
Comment (hedasilv) [inline server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:193]: Possibly a silly question - why is this part needed? `...Object.values(fullTreeBlobs).map(async (blob) => blob)`
Wouldn't it be enough to do `blobPs.push(async () => { fullTreeBlobs });`?

File: server/gitrest/packages/gitrest-base/src/utils/filesystems.ts [hunk 15]
```diff
@@ -3,11 +3,18 @@
...
 
 export class NodeFsManagerFactory implements IFileSystemManagerFactory {
     public create(params?: IFileSystemManagerParams): IFileSystemManager {
         return fs;
     }
 }
+
+export class MemFsManagerFactory implements IFileSystemManagerFactory {
+    public create(params?: IFileSystemManagerParams): IFileSystemManager {
+        return (new Volume() as unknown) as IFileSystemManager;
```
[the remark this answers is not in the document]
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/filesystems.ts:18]: Hey Xin, this is only used very briefly in 1 function call and should be picked up by GC like any other variable. It is not stored external to the function, and is identical to the logic in Spork, which had a memory leak because of a global object

**Comment by znewton** [comment]
Merging because confident in correctness (tested with R11s and FRS extensively). Also, new functionality is hidden behind a feature flag, so it will only be enabled in intentional scenarios.

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts [hunk 16]
```diff
@@ -101,45 +141,83 @@ function getSummaryObjectFromWholeSummaryTreeEntry(entry: WholeSummaryTreeEntry)
...
+                    depth + 1,
+                );
+                const baseTreeEntryPath = treeEntry.path.replace(fullTreePath, "");
+                treeEntries.push(...builtFullGitTree.tree.tree.map((fullTreeEntry => ({
+                    ...fullTreeEntry,
+                    path: `${baseTreeEntryPath}${fullTreeEntry.path}`,
+                }))));
+                const fullTreeBlobs = {
+                    ...fullTree.blobs,
+                    ...builtFullGitTree.blobs,
```
Comment (hedasilv) [inline server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:191]: Just to make sure that my mental model is correct:
1. `fullTree.blobs` are the top level blobs, included at the full tree (encoded as a single blob), depth === 0
2. `builtFullGitTree.blobs` would include the other possible blobs at depths === 1+. Those would be obtaining by using the top level tree (the blobs were already taken care of in step 1 above) to recursively obtain those blobs. 

Is that correct? The part that confused me a bit is what the tree would look like when in the recursion loop. This comment said:
````
// All blobs associated with full git tree are included in the full git tree blob, and
```
Does that mean that the `fullTree.tree` that is passed to `buildFullGitTreeFromGitTree` will also have the `fullTreePath` encoded in its sha?

**Review by hedasilv — APPROVED** [review]

File: server/gitrest/packages/gitrest-base/src/routes/summaries.ts (hunk 12 above)
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/routes/summaries.ts:122]: Discussed offline that `enableLowIoWrite` is added to metrics internally within `GitWholeSummaryManager` which is the only place that enableLowIoWrite alters logic. Can follow up with a PR to hoist it though

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts (hunk 13 above)
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:187]: Correct. If the path is `".app/.fullTree"`, `baseTreeEntryPath` will equal `".app/"`. If path is just `.fullTree`, then `baseTreeEntryPath` will be `""`

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts (hunk 14 above)
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:193]: `fullTreeBlobs` is type `Record<string, IBlob>` and `blobPs` is type `Promise<IBlob>[]`.

`blobPs.push(async () => ({ fullTreeBlobs }));` would push type `Promise<{ fullTreeBlobs: Record<string, IBlob> }>` to a `Promise<IBlob>[]` type, which isn't compatible and wouldn't be consumable in the same way as the other blobs promises

Comment (hedasilv) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:193]: Ah, got it. I misunderstood the types

File: server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts (hunk 16 above)
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:191]: > Just to make sure that my mental model is correct:
> - `fullTree.blobs` are the top level blobs, included at the full tree (encoded as a single blob), depth === 0
> - `builtFullGitTree.blobs` would include the other possible blobs at depths === 1+. Those would be obtaining by using the top level tree (the blobs were already taken care of in step 1 above) to recursively obtain those blobs.

This is almost correct. The part that's missing is when `fullTree.blobs` contains a fullTree single blob in itself, in which case we need to parse out and unpack that internal fullTree single blob, which will give us a new list of blobs that were hidden within the internal fullTreeSingle blob. Only the internal fullTree single blob is contained in the fullTree.blobs, not the internal fullTree's internal blobs. We pass `false, /* retrieveBlobs */` into the recursive `buildFullGitTreeFromGitTree` calls because we know that any blobs not contained in an internal fullGitTree blob are contained in the current fullGitTree blob.

I might not be explaining that well, but hopefully it kinda makes sense.

Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts:191]: > Does that mean that the `fullTree.tree` that is passed to `buildFullGitTreeFromGitTree` will also have the `fullTreePath` encoded in its sha?

This is a different question entirely, because there are a lot of reasons that the `fullGitTree-Blob` sha would be different from the `fullGitTree` original tree sha. A lot of the complexity of this PR is due to this being the case, and there's not really a way to workaround it (that I could think of). That's OK though, because "unpacking" and writing the "packed" git tree back into Git (in-memory) as a normal Git Tree results in the same sha as the original git tree that was "packed".
